src/main/java/io/confluent/idesidecar/restapi/cache/Clients.java

```diff
@@ -0,0 +1,129 @@
+import com.github.benmanes.caffeine.cache.RemovalListener;
+import io.confluent.idesidecar.restapi.connections.ConnectionState;
+import io.confluent.idesidecar.restapi.events.Lifecycle;
+import io.quarkus.logging.Log;
+import jakarta.enterprise.event.ObservesAsync;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Supplier;
+
+/**
+ * Utilities to obtain and cache clients for a given connection and client ID.
+ */
+public abstract class Clients<T extends AutoCloseable> {
+
+  /**
+   * Store an instance of a Caffeine cache for each connection. The cache will store
+   * clients by client ID and its policy may be configured by the {@link #getCaffeineSpec()}
+   * method.
+   */
+  private final Map<String, Cache<String, T>> clientsByIdByConnections =
+      new ConcurrentHashMap<>();
+
+  RemovalListener<String, T> removalListener = (key, value, cause) -> {
+    try {
+      if (value != null) {
+        Log.debugf("Closing client %s", value);
+        value.close();
+      }
+    } catch (Throwable t) {
+      // Ignore these as we don't care
+    }
+  };
+
+  /**
+   * Get a client for the given connection and client ID. If the client does not
+   * already exist, it will be created using the provided factory.
+   *
+   * @param connectionId     the ID of the connection
+   * @param clientId         the identifier of the client
+   * @param factory          the method that will create the client if there is not already one
+   * @return the client
+   */
+  public T getClient(
+      String connectionId,
+      String clientId,
+      Supplier<T> factory
+  ) {
+    return clientsForConnection(connectionId).asMap().computeIfAbsent(
+        clientId,
+        k -> factory.get()
+    );
+  }
+
+  private Cache<String, T> clientsForConnection(String connectionId) {
+    return clientsByIdByConnections.computeIfAbsent(connectionId, k -> createCache());
+  }
+
+  int clientCount() {
+    return clientsByIdByConnections
+        .values()
+        .stream()
+        .map(Cache::asMap)
+        .map(Map::size)
+        .mapToInt(Integer::intValue)
+        .sum();
+  }
+
+  int clientCount(String connectionId) {
+    return clientsForConnection(connectionId).asMap().size();
+  }
+
+  void clearClients(String connectionId) {
+    var oldCache = clientsByIdByConnections.put(connectionId, createCache());
+    if (oldCache != null) {
+      // Invalidation will trigger the removal listener, which will close the clients
+      oldCache.invalidateAll();
+    }
+  }
+
+  /**
+   * Get the Caffeine spec to use for the caches. By default, this will be an empty spec.
+   * See the
+   * <a href="https://github.com/ben-manes/caffeine/wiki/Specification">Caffeine Spec</a>
+   * for more information on the format. Inherited classes may override this method
+   * and return a different spec.
+   * @return the Caffeine spec
+   */
+  protected CaffeineSpec getCaffeineSpec() {
+    return CaffeineSpec.parse("");
+  }
```

Comment:
The implementation of `getCaffeineSpec()` has no parameters and there's no `Clients` state, which means in practice the specs will always be the same. Is there a benefit to defining the spec with a method, versus passing the spec into the `Clients` constructor as an immutable spec for use in all embedded caches?

Reply:
Makes sense, thanks! I'll incorporate that.